examples/frompapers/Brunel_2000.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+"""
+Fig. 8 from:
+
+Brunel, N. Dynamics of Sparsely Connected Networks of Excitatory and
+Inhibitory Spiking Neurons. J Comput Neurosci 8, 183–208
+(2000). https://doi.org/10.1023/A:1008925309027
+
+Inspired by http://neuronaldynamics.epfl.ch
+
+Sebastian Schmitt, 2022
+"""
+
+import random
+from brian2 import *
+import matplotlib.pyplot as plt
+
+
+def sim(g, nu_ext_over_nu_thr, sim_time, ax_spikes, ax_rates, rate_tick_step):
+    """
+    g -- relative inhibitory to excitatory synaptic strength
+    nu_ext_over_nu_thr -- ratio of external stimulus rate to threshold rate
+    sim_time -- simulation time
+    ax_spikes -- matplotlib axes to plot spikes on
+    ax_rates -- matplotlib axes to plot rates on
+    rate_tick_step -- step size for rate axis ticks
+    """
+
+    # network parameters
+    N_E = 10000
+    gamma = 0.25
+    N_I = round(gamma * N_E)
+    N = N_E + N_I
+    epsilon = 0.1
+    C_E = epsilon * N_E
+    C_ext = C_E
+
+    # neuron parameters
+    tau = 20 * ms
+    theta = 20 * mV
+    V_r = 10 * mV
+    tau_rp = 2 * ms
+
+    # synapse parameters
+    J = 0.1 * mV
+    D = 1.5 * ms
+
+    # external stimulus
+    nu_thr = theta / (J * C_E * tau)
+
+    defaultclock.dt = 0.1 * ms
+
+    neurons = NeuronGroup(N,
+                          """
+                          dv/dt = -v/tau : volt (unless refractory)
+                          index : integer (constant)
+                          """,
+                          threshold="v > theta",
+                          reset="v = V_r",
+                          refractory=tau_rp,
+                          method="exact",
+    )
+
+    excitatory_neurons = neurons[:N_E]
+    inhibitory_neurons = neurons[N_E:]
+
+    exc_synapses = Synapses(excitatory_neurons, target=neurons, on_pre="v += J", delay=D)
+    exc_synapses.connect(p=epsilon)
+
+    inhib_synapses = Synapses(inhibitory_neurons, target=neurons, on_pre="v += -g*J", delay=D)
+    inhib_synapses.connect(p=epsilon)
+
+    nu_ext = nu_ext_over_nu_thr * nu_thr
+
+    external_poisson_input = PoissonInput(
+        target=neurons, target_var="v", N=C_ext, rate=nu_ext, weight=J
+    )
+
+    rate_monitor = PopulationRateMonitor(neurons)
+
+    # shuffle indices to randomly select neurons
```

Comment:
I need to select a random subset because the first 10k neurons are exc. and the rest (2.5k) is inh. I want to select 50 neurons across all 12.5k neurons.

I was trying to follow https://brian2.readthedocs.io/en/stable/user/recording.html#recording-random-subsets-of-neurons.

I thought that `.index` is a special variable (at least it does already exist) that steers the order for contiguous slicing.

What's the correct idiom to record spikes from a random subset?

Reply:
Ah I see, now I get it. At the moment, you cannot currently record random subsets of neurons with a `SpikeMonitor`. The first paragraph of the documentation page you link to mentions that, but I see that the code example with `index` later is misleading. The idea in that example is to use `index` instead of the places where you'd normally use `i` in your assignments/connection statements, etc. So instead of recording a random subset from an ordered population, you'd record a contiguous subset from a shuffled population. In the context of the Brunel example, it would mean to define the connections like this:
```Python
exc_synapses = Synapses(neurons, neurons, on_pre="v += J", delay=D)
exc_synapses.connect('index < N_E', p=epsilon)
inh_synapses = Synapses(neurons, neurons, on_pre="v += -g*J", delay=D)
inh_synapses.connect('index >= N_E', p=epsilon)
```
But I don't really think this is necessary here. As you inadvertently demonstrated, just selecting excitatory gives you figures that seem to nicely match the figures in the paper. I am actually not sure that the paper really plots random neurons and not random excitatory neurons. It is quite common in these networks to only plot excitatory neurons, and other implementations, e.g. the one in [PyNEST](https://github.com/nest/nest-simulator/blob/master/pynest/examples/brunel_delta_nest.py) certainly do.

> I thought that `.index` is a special variable (at least it does already exist) that steers the order for contiguous slicing.

No, it isn't a special variable. It only exists because you define it as part of your equations :blush: 

> What's the correct idiom to record spikes from a random subset?

As mentioned above, this doesn't currently work. This will soon change with PR #1229, though.

Reply:
Thanks!

Reply:
You're welcome. Could you please also remove the `index : integer (constant)` in the equations? Thanks!

Reply:
Done